q2_composition/_ancombc2.py

```diff
@@ -0,0 +1,782 @@
+        with their reference levels as explained above.
+    '''
+    # restructure categorical variable columns into 'variable::level' format
+    for slice_df in slices.values():
+        for slice_column in slice_df.columns:
+            if _is_categorical(slice_column, metadata):
+                variable, level = _parse_variable_and_level(
+                    slice_column, metadata
+                )
+                slice_df.rename(
+                    {slice_column: variable + '::' + level},
+                    axis='columns',
+                    inplace=True
+                )
+
+    # deduce reference level of each categorical variable and annotate columns
+    # any slice will do except for structural zeros
+    reference_levels = _deduce_reference_levels(slices['lfc'], metadata)
+    for slice_df in slices.values():
+        for column in slice_df.columns:
+            if _is_categorical(column, metadata):
+                variable, level = _parse_variable_and_level(column, metadata)
+                reference_level = reference_levels[variable]
+                slice_df[column].attrs['variable'] = variable
+                slice_df[column].attrs['level'] = level
+                slice_df[column].attrs['reference'] = reference_level
+
+    return slices
+
+
+def _is_categorical(column: str, metadata: qiime2.Metadata) -> bool:
+    '''
+    Determines whether `column` refers to a categorical variable.
+
+    Parameters
+    ----------
+    column : str
+        The column name as returned by ANCOMBC2. If in reference to a
+        categorical column then a level will be appended to the variable name.
+    metadata : qiime2.Metadata
+        The per-sample metadata.
+
+    Returns
+    -------
+    bool
+        Whether `column` refers to a categorical variable.
+    '''
+    # reverse sort to handle md columns that are prefixes of other md columns
+    for md_column in sorted(list(metadata.columns), reverse=True):
+        if column.startswith(md_column):
+            if isinstance(
+                metadata.get_column(md_column), CategoricalMetadataColumn
+            ):
+                return True
+
+            return False
+
+    return False
+
+
+def _parse_variable_and_level(
+    column: str, metadata: qiime2.Metadata
+) -> tuple[str, str]:
+    '''
+    Parses `column` into its variable name and level name components. Assumes
+    that `column` refers to a categorical variable.
+
+    Parameters
+    ----------
+    column : str
+        The model statistics column name that is a concatenation of a variable
+        name and a level name.
+    metadata : qiime2.Metadata
+        The per-sample metadata.
+
+    Returns
+    -------
+    tuple[str, str]
+        The variable name and the level name.
+
+    Raises
+    ------
+    ValueError
+        If `column` does not refer to categorical variable.
+    ValueError
+        If the categorical variable
```

Comment:
```suggestion
        If the categorical variable referenced by `column`
        was not found in the metadata.
```
I'm assuming this was what was intended here?